File: src/lua/flux_constraint.lua
-- FLUX Constraint Engine — Lua 5.3+
-- Pure INT8 saturated constraint checking. Zero dependencies.

local FluxConstraint = {}
FluxConstraint.__index = FluxConstraint

FluxConstraint.INT8_MIN = -127
FluxConstraint.INT8_MAX = 127

function FluxConstraint.saturate(val)
    return math.max(-127, math.min(127, val))
end

function FluxConstraint.new(constraints)
    assert(type(constraints) == "table" and #constraints > 0, "Non-empty constraints required")
    assert(#constraints <= 8, "Max 8 constraints")
    local self = setmetatable({}, FluxConstraint)
    self.constraints = {}
    for i, c in ipairs(constraints) do
        self.constraints[i] = {
            lo = FluxConstraint.saturate(c.lo or c[1]),
            hi = FluxConstraint.saturate(c.hi or c[2]),
            name = c.name or ("C" .. i)
        }
    end
    return self
end

function FluxConstraint:check(value)
    local val = FluxConstraint.saturate(value)
    local errorMask, violatedLo, violatedHi, violatedCount = 0, 0, 0, 0
    local details = {}

    for i, c in ipairs(self.constraints) do
        local loFail = val < c.lo
        local hiFail = val > c.hi
        local passed = not loFail and not hiFail

        if not passed then
            errorMask = errorMask | (1 << (i - 1))
            violatedCount = violatedCount + 1
        end
        if loFail then violatedLo = violatedLo | (1 << (i - 1)) end
        if hiFail then violatedHi = violatedHi | (1 << (i - 1)) end

        details[i] = { name = c.name, lo = c.lo, hi = c.hi, passed = passed }
    end

    local nc = #self.constraints
    local severity = 0
    if violatedCount > 0 then
        if violatedCount <= nc // 4 then severity = 1
        elseif violatedCount <= nc // 2 then severity = 2
        else severity = 3 end
    end

    return {
        errorMask = errorMask,
        severity = severity,
        violatedLo = violatedLo,
        violatedHi = violatedHi,
        violatedCount = violatedCount,
        passed = violatedCount == 0,
        details = details
    }
end

function FluxConstraint:checkBatch(values)
    local results, stats = {}, { pass = 0, caution = 0, warning = 0, critical = 0 }
    local labels = { [0] = "pass", [1] = "caution", [2] = "warning", [3] = "critical" }
    for i, v in ipairs(values) do
        local r = self:check(v)
        results[i] = r
        stats[labels[r.severity]] = stats[labels[r.severity]] + 1
    end
    return results, stats
end

function FluxConstraint:benchmark(iterations)
    iterations = iterations or 1000000
    local t0 = os.clock()
    for i = 1, iterations do
        self:check((i % 254) - 127)
    end
    local elapsed = os.clock() - t0
    local rate = iterations * #self.constraints / elapsed
    return rate, elapsed * 1000
end

-- Presets
FluxConstraint.presets = {
    aviation = {
        { lo = -55, hi = 70, name = "cabin_temp_C" },
        { lo = 75, hi = 101, name = "cabin_pressure_kPa" },
        { lo = 0, hi = 100, name = "fuel_flow_pct" },
        { lo = 60, hi = 100, name = "hydraulic_pct" },
    },
    medical = {
        { lo = 36, hi = 38, name = "body_temp_C" },
        { lo = 60, hi = 100, name = "heart_rate_bpm" },
        { lo = 95, hi = 100, name = "spo2_pct" },
        { lo = 80, hi = 120, name = "bp_systolic_mmHg" },
    },
    maritime = {
        { lo = -2, hi = 35, name = "sea_temp_C" },
        { lo = 50, hi = 100, name = "hull_integrity_pct" },
        { lo = 0, hi = 50, name = "wave_height_m" },
        { lo = 0, hi = 80, name = "wind_speed_kn" },
    },
    automotive = {
        { lo = -40, hi = 60, name = "battery_temp_C" },
        { lo = 0, hi = 100, name = "soc_pct" },
        { lo = 0, hi = 100, name = "charge_rate_pct" },
        { lo = 20, hi = 80, name = "cabin_temp_C" },
    },
    energy = {
        { lo = 49, hi = 51, name = "grid_freq_Hz_x10" },
        { lo = 95, hi = 105, name = "voltage_pct" },
        { lo = 0, hi = 80, name = "transformer_temp_C" },
        { lo = 0, hi = 100, name = "line_load_pct" },
    },
}

function FluxConstraint.fromPreset(name)
    local cs = FluxConstraint.presets[name]
    if not cs then error("Unknown preset: " .. name) end
    return FluxConstraint.new(cs)
end

-- Self-test
if arg and arg[0] and arg[0]:match("flux_constraint") then
    print("FLUX Constraint Engine - Lua")
    print("============================")

    assert(FluxConstraint.saturate(-128) == -127, "sat(-128)")
    assert(FluxConstraint.saturate(128) == 127, "sat(128)")
    print("  saturate: OK")

    local fc = FluxConstraint.new({ { lo = 0, hi = 100, name = "test" } })
    assert(fc:check(50).passed, "pass")
    assert(not fc:check(150).passed, "fail")
    print("  check: OK")

    local fc2 = FluxConstraint.new({
        { lo = 0, hi = 10, name = "a" }, { lo = 0, hi = 10, name = "b" },
        { lo = 0, hi = 10, name = "c" }, { lo = 0, hi = 10, name = "d" }
    })
    assert(fc2:check(50).severity == 3, "critical")
    print("  severity: OK")

    local fc3 = FluxConstraint.fromPreset("aviation")
    assert(#fc3.constraints == 4, "preset")
    print("  presets: OK")

    local rate, ms = fc3:benchmark()
    print(string.format("  Benchmark: %.1fM checks/sec (%.1fms)", rate / 1e6, ms))

    print("  All tests pass")
end

return FluxConstraint

File: src/csharp/FluxConstraint.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;

namespace Flux
{
    public enum Severity { Pass = 0, Caution = 1, Warning = 2, Critical = 3 }

    public class ConstraintDef
    {
        public int Lo { get; set; }
        public int Hi { get; set; }
        public string Name { get; set; }

        public ConstraintDef(int lo, int hi, string name)
        {
            Lo = Saturate(lo);
            Hi = Saturate(hi);
            Name = name;
        }
    }

    public class FluxResult
    {
        public int ErrorMask { get; set; }
        public Severity Severity { get; set; }
        public int ViolatedLo { get; set; }
        public int ViolatedHi { get; set; }
        public int ViolatedCount { get; set; }
        public bool Passed => Severity == Severity.Pass;
        public List<(string Name, int Lo, int Hi, bool Passed)> Details { get; set; } = new();
    }

    public class FluxChecker
    {
        private readonly List<ConstraintDef> _constraints;

        public FluxChecker(List<ConstraintDef> constraints)
        {
            if (constraints == null || constraints.Count == 0)
                throw new ArgumentException("Non-empty constraints required");
            if (constraints.Count > 8)
                throw new ArgumentException("Max 8 constraints (INT8 × 8 flat bounds)");
            _constraints = constraints;
        }

        public FluxResult Check(int value)
        {
            int val = Saturate(value);
            var result = new FluxResult();
            int violated = 0;

            for (int i = 0; i < _constraints.Count; i++)
            {
                var c = _constraints[i];
                bool loFail = val < c.Lo;
                bool hiFail = val > c.Hi;
                bool passed = !loFail && !hiFail;

                if (!passed) { result.ErrorMask |= (1 << i); violated++; }
                if (loFail) result.ViolatedLo |= (1 << i);
                if (hiFail) result.ViolatedHi |= (1 << i);

                result.Details.Add((c.Name, c.Lo, c.Hi, passed));
            }

            int nc = _constraints.Count;
            result.ViolatedCount = violated;
            result.Severity = violated == 0 ? Severity.Pass
                : violated <= nc / 4 ? Severity.Caution
                : violated <= nc / 2 ? Severity.Warning
                : Severity.Critical;

            return result;
        }

        public (List<FluxResult> Results, Dictionary<string, int> Stats) CheckBatch(IEnumerable<int> values)
        {
            var results = new List<FluxResult>();
            var stats = new Dictionary<string, int>
            {
                ["pass"] = 0, ["caution"] = 0, ["warning"] = 0, ["critical"] = 0
            };

            foreach (var v in values)
            {
                var r = Check(v);
                results.Add(r);
                stats[r.Severity.ToString().ToLower()]++;
            }
            return (results, stats);
        }

        public (double Rate, double Ms) Benchmark(int iterations = 1_000_000)
        {
            var sw = Stopwatch.StartNew();
            for (int i = 0; i < iterations; i++)
                Check((i % 254) - 127);
            sw.Stop();
            double rate = iterations * _constraints.Count / (sw.Elapsed.TotalMilliseconds / 1000.0);
            return (rate, sw.Elapsed.TotalMilliseconds);
        }

        public static FluxChecker FromPreset(string name)
        {
            if (!Presets.ContainsKey(name))
                throw new ArgumentException($"Unknown preset: {name}. Available: {string.Join(", ", Presets.Keys)}");
            return new FluxChecker(Presets[name]);
        }

        public static int Saturate(int val) => Math.Max(-127, Math.Min(127, val));

        public static readonly Dictionary<string, List<ConstraintDef>> Presets = new()
        {
            ["aviation"] = new()
            {
                new(-55, 70, "cabin_temp_C"),
                new(75, 101, "cabin_pressure_kPa"),
                new(0, 100, "fuel_flow_pct"),
                new(60, 100, "hydraulic_pct")
            },
            ["medical"] = new()
            {
                new(36, 38, "body_temp_C"),
                new(60, 100, "heart_rate_bpm"),
                new(95, 100, "spo2_pct"),
                new(80, 120, "bp_systolic_mmHg")
            },
            ["maritime"] = new()
            {
                new(-2, 35, "sea_temp_C"),
                new(50, 100, "hull_integrity_pct"),
                new(0, 50, "wave_height_m"),
                new(0, 80, "wind_speed_kn")
            },
            ["automotive"] = new()
            {
                new(-40, 60, "battery_temp_C"),
                new(0, 100, "soc_pct"),
                new(0, 100, "charge_rate_pct"),
                new(20, 80, "cabin_temp_C")
            },
            ["energy"] = new()
            {
                new(49, 51, "grid_freq_Hz_x10"),
                new(95, 105, "voltage_pct"),
                new(0, 80, "transformer_temp_C"),
                new(0, 100, "line_load_pct")
            }
        };
    }

    // Tests
    public static class FluxTests
    {
        public static void RunAll()
        {
            Console.WriteLine("╔══════════════════════════════════════════════════════╗");
            Console.WriteLine("║  FLUX Constraint Engine — C#                         ║");
            Console.WriteLine("╚══════════════════════════════════════════════════════╝\n");

            // Test 1: Saturate
            Assert(FluxChecker.Saturate(-128) == -127, "saturate(-128)");
            Assert(FluxChecker.Saturate(128) == 127, "saturate(128)");
            Assert(FluxChecker.Saturate(0) == 0, "saturate(0)");
            Console.WriteLine("✓ saturate boundaries");

            // Test 2: Pass
            var fc = new FluxChecker(new List<ConstraintDef> { new(0, 100, "test") });
            var r1 = fc.Check(50);
            Assert(r1.Passed, "should pass");
            Console.WriteLine("✓ single constraint pass");

            // Test 3: Fail
            var r2 = fc.Check(150);
            Assert(!r2.Passed, "should fail");
            Assert(r2.ErrorMask != 0, "should have error");
            Console.WriteLine("✓ single constraint fail");

            // Test 4: Multi
            var fc2 = new FluxChecker(new List<ConstraintDef>
            {
                new(0, 50, "a"), new(0, 100, "b"), new(-10, 10, "c")
            });
            var r3 = fc2.Check(30);
            Assert(!r3.Passed, "should fail");
            Console.WriteLine("✓ multi constraint mixed");

            // Test 5: Severity
            var fc3 = new FluxChecker(new List<ConstraintDef>
            {
                new(0, 10, "a"), new(0, 10, "b"), new(0, 10, "c"), new(0, 10, "d")
            });
            var r4 = fc3.Check(50);
            Assert(r4.Severity == Severity.Critical, "should be critical");
            Assert(r4.ViolatedCount == 4, "all 4 should fail");
            Console.WriteLine("✓ severity critical");

            // Test 6: Preset
            var fc4 = FluxChecker.FromPreset("aviation");
            Assert(fc4._constraints.Count == 4, "aviation should have 4");
            Console.WriteLine("✓ preset loading");

            // Test 7: Batch
            var (results, stats) = fc.CheckBatch(new[] { -60, 0, 50, 100, 127 });
            Assert(results.Count == 5, "should have 5 results");
            Console.WriteLine("✓ batch checking");

            // Benchmark
            var (rate, ms) = fc4.Benchmark();
            Console.WriteLine($"\n  Benchmark: {rate / 1e6:F1}M checks/sec ({ms:F1}ms)");
            Console.WriteLine("\n  ✓ All tests pass");
        }

        static void Assert(bool condition, string msg)
        {
            if (!condition) throw new Exception($"FAIL: {msg}");
        }
    }
}
